TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/TrnRequestMetadataBackfillJob.cs

```diff
@@ -0,0 +1,358 @@
+                g => (g.First().TrnRequestId)
+            );
+
+        var contactsQuery = new QueryExpression(Contact.EntityLogicalName)
+        {
+            ColumnSet = new ColumnSet(
+                Contact.Fields.CreatedOn,
+                Contact.Fields.dfeta_TrnRequestID
+            ),
+            PageInfo = new()
+            {
+                PageNumber = 1,
+                Count = 500
+            }
+        };
+        contactsQuery.Criteria.AddCondition(Contact.Fields.dfeta_TrnRequestID, ConditionOperator.NotNull);
+        EntityCollection result;
+
+        //fetch contacts from crm that have a trnrequestid
+        do
+        {
+            result = await serviceClient.RetrieveMultipleAsync(contactsQuery);
+            foreach (var record in result.Entities)
+            {
+                teacherIds.Add(record.Id, record.GetAttributeValue<string>(Contact.Fields.dfeta_TrnRequestID));
+            }
+            contactsQuery.PageInfo.PageNumber++;
+            contactsQuery.PageInfo.PagingCookie = result.PagingCookie;
+        } while (result.MoreRecords);
+
+        var connstring = config.GetPostgresConnectionString();
+        using var conn = new NpgsqlConnection(connstring);
+        await conn.OpenAsync();
+        using var transaction = await conn.BeginTransactionAsync();
+        await CreateTempTableAsync(conn, transaction);
+        try
+        {
+            var totalBatches = (int)Math.Ceiling((double)teacherIds.Count() / BATCH_SIZE);
+            for (int batch = 0; batch < totalBatches; batch++)
+            {
+                var contactsToSync = teacherIds.Skip(batch * BATCH_SIZE).Take(BATCH_SIZE).ToList();
+                var recordsToInsert = new List<TrnRequestMetadata>();
+
+                foreach (var contact in contactsToSync)
+                {
+                    //fetch audit records
+                    var query = new QueryExpression(Audit.EntityLogicalName)
+                    {
+                        ColumnSet = new ColumnSet(true),
+                        Orders =
+                    {
+                        new OrderExpression(Audit.Fields.CreatedOn, OrderType.Ascending)
+                    },
+                        PageInfo = new()
+                        {
+                            PageNumber = 1,
+                            Count = 500
+                        }
+                    };
+                    query.Criteria.AddCondition(Audit.Fields.ObjectId, ConditionOperator.Equal, contact.Key);
+                    query.Criteria.AddCondition(Audit.Fields.ObjectTypeCode, ConditionOperator.Equal, "contact");
+                    query.Criteria.AddCondition(Audit.Fields.Operation, ConditionOperator.Equal, 1); //create
+                    var auditRecords = await serviceClient.RetrieveMultipleAsync(query);
+                    var audit = auditRecords.Entities.FirstOrDefault();
+
+                    //audit is stored in json
+                    if (audit != null)
```

Comment:
Do we need to deserialize manually here? We're using strong types in `TrsDataSyncHelper`.